Sourcery/Models/TemplateContext.swift

```diff
@@ -118,32 +118,60 @@ public final class Types: NSObject, SourceryModel {
     // sourcery: skipDescription, skipEquality, skipCoding
     /// Types based on any other type, grouped by its name, even if they are not known.
     /// `types.based.MyType` returns list of types based on `MyType`
-    public lazy internal(set) var based: BasedTypesCollection = {
-        BasedTypesCollection(types: self.types)
+    public lazy internal(set) var based: TypesCollection = {
+        TypesCollection(
+            types: self.types,
+            collection: { Array($0.based.keys) },
+            valueForKey: { (collection: TypesCollection, key: String) -> Any? in
+                guard collection.all.first(where: { $0.name == key }) != nil else {
+                    throw "Unknown type \(key)"
+                }
+                return collection.types[key]
+        })
     }()
 
     // sourcery: skipDescription, skipEquality, skipCoding
     /// Classes inheriting from any known class, grouped by its name.
     /// `types.inheriting.MyClass` returns list of types inheriting from `MyClass`
-    public lazy internal(set) var inheriting: InheritingTypesCollection = {
-        InheritingTypesCollection(types: self.types)
+    public lazy internal(set) var inheriting: TypesCollection = {
+        TypesCollection(
+            types: self.types,
+            collection: { Array($0.inherits.keys) },
+            valueForKey: { (collection: TypesCollection, key: String) -> Any? in
```

Comment:
this valueForKey still is the same across all variants, so why not just keep it in the base class rather than parametrize?